### ISSUE TITLE
After disabling and enabling AD using a different user for test and enable auth , this user is not allowed to log in to rancher server anymore. 

### ISSUE THREAD

**Issue by sangeethah** [rancher/rancher#13666]
**Rancher versions: v2.0.1-rc6
**Steps to Reproduce:**

Enable AD using user - "testuser1" for test and enable auth.
Add "testuser2"in the "Site Access" list - "Allow members of Clusters, Projects, plus Authorized Users and Organizations".

Make sure that both "testuser1 and "testuser2" are allowd to access rancher-server.

Disable AD auth.
Enable AD auth using user - "testuser2" for test and enable auth.

Try to log in as "testuser2". Log in fails with following errors seen in logs:

"2018/05/23 16:33:21 [ERROR] API error response 500 for POST /v3-public/activeDirectoryProviders/activedirectory?action=login. Cause: can't find unique user for principal activedirectory_user://CN=<cn>,CN=Users,DC=<dc1>,DC=<dc2>,DC=<dc3>"

**Comment by maxkorp** [comment]
CC @cjellick @deniseschannon 

So the flow here is kind of odd but I've found where it dies.

- Create testuser2. This creates `u-abc` with no AD principal.
- Enable AD.
- Login to testuser2 via AD. This creates `u-xyz` with the AD principal `AD-testuser2`.
- Disable AD.
- Login with testuser2 (`u-abc`).
- Enable AD using the `testuser2` as the AD account. This associates the AD principal `AD-testuser2` with `u-abc`. 

At this point, both `u-abc` and `u-xyz` are associated with the same AD principal, which is where the login fails.

I have a few solutions proposed below. I think the core decision factor here is if  users with the same user name created pre auth, and then logged in via AD are intentionally separate users, or if they really should be bound together. I don't know if that's a constant prescription we make one way or the other, or if that varies from user/install to user/install. 

Potential solutions, including some which may work in concert:
- When you login with AD, if no user already exists with that primary, attempt to automatically bind to an existing local user with the same username (if the user is able to log in). 
- Same as above, but give the user the option to make the binding to the found user
- Some way to manually merge existing accounts or add/remove/transfer principals. I know this something already being talked about, but it directly impacts this bug. 
- When activating AD, if another user is found with the same principal, rather than just blindly binding the principal to the current user, allow the user to remove the principal from the old user (would have to enforce a password set on local login, as well as potentially generating a local principal or login name or whatever), or to log in to AD as a different user. Generally, disallow getting the principal into a bad state.
- Nuclear option of deleting all AD principals when AD is deactivated, but if it's being reactivated, we'd need some way to reattach rather than generating all new accounts again.

I'm going to try testing with another provider or two to see if the behavior is the same there.

**Comment by cjellick** [comment]
We marked this as a p0 out of a concern that the impact was that no user could log in. @maxkorp investigated and found it was limited to just the AD account associated with the user who turned on AD the second time. As this is fairly limited in impact, moving to p1.

**Comment by cjellick** [comment]
moving this out of this release, as it is a minor issue

**Comment by sangeethah** [comment]
This issue also reproduces in this use case:

Enable github auth (as githubuser1) when logged in as Default Admin user.
Disable github auth.
Create a new admin user say - "admin1"
Enable github auth  (as githubuser1) when logged in as "admin1".
Log out
Try to log in using github as githubuser1.

User is presented with the following error message:
```
2018/09/14 19:29:24 [ERROR] API error response 500 for POST /v3-public/githubProviders/github?action=login. Cause: can't find unique user for principal github_user://4266958
```

**Comment by maxkorp** [comment]
I suspect this is going to occur with all auth providers, as a new user is created for the second user's login with the auth provider, which is not associated with their built in account, and then you end up with 2 users with that base login name principal

**Comment by mrajashree** [comment]
Yes, maybe we have to check if it's possible to remove external auth principals from all users if auth is being disabled

**Comment by jfrederickson** [comment]
I just ran into what appears to be this issue after a similar series of actions with a SAML account on Rancher 2.2.6; I don't think it should be closed.

```
2019/10/02 22:03:09 [ERROR] SAML: Failed getting user with error: can't find unique user for principal adfs_user://jfrederickson
```

**Comment by oskapt** [comment]
@jfrederickson since this issue is closed, i'd recommend that you open a new issue and reference this one. new issues are more likely to be seen.

**Comment by nickngch** [comment]
Encountered this issue on Rancher 2.3.3

Followed https://github.com/rancher/rancher/issues/13666#issuecomment-421466497 to reproduced the issue in my environment.

Workaround: 
1. search the user from `kubectl get users.management.cattle.io -o yaml `
2. Remove the user based on the display name

**Comment by amagura** [comment]
This is literally what's happening to me right now.  #29433

**Comment by vroetman** [comment]
I just ran into this with Rancher 2.4.12 (Docker). Deleting the second user with kubectl worked for me.

**Comment by dnoland1** [comment]
Hit this today on a new system with only a single non-admin user:
```
$ kubectl get users -o custom-columns=NAME:metadata.name,USERNAME:username,DISPLAY_NAME:displayName,PRINCIPAL_IDS:principalIds,CREATE_DATE:metadata.creationTimestamp --sort-by metadata.name | grep -v "System account"
NAME           USERNAME   DISPLAY_NAME                         PRINCIPAL_IDS                                                                       CREATE_DATE
u-ikdhxu5efy   <none>     redacted                             [azuread_user://5xxxxxxx-xxxx-xxxxx-xxxx-xxxxxxxxxxxe local://u-ikdhxu5efy]         2021-03-03T22:22:38Z
user-zrntg     admin      Default Admin                        [local://user-zrntg azuread_user://5xxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxe]            2021-03-03T21:12:59Z
```

**Comment by dnoland1** [comment]
Also hitting this on Okta as well with Rancher v2.5.6:
```
NAME           USERNAME                DISPLAY_NAME                            PRINCIPAL_IDS                                                              CREATE_DATE
u-kkwxnqsrzz   <none>                  redacted                                [okta_user://xxxxxxxxxx@xxxxxxxxxxx.com local://u-kkwxnqsrzz]              2021-01-06T01:09:12Z
user-bx22k     admin                   Default Admin                           [local://user-bx22k okta_user://xxxxxxxxxx@xxxxxxxxxxx.com]                2021-01-06T01:09:19Z
```

**Comment by ccckblaze** [comment]
+1
keycloak have the same issue

**Comment by paynejacob** [comment]
related issue https://github.com/rancher/rancher/issues/33331

**Comment by paynejacob** [comment]
After talking to @dramich and @vincent99 the solution to this is to return a 4XX response when a principal is going to be bound to 2 users. The user should be sent to documentation for why this is not possible.

When a user configures an authentication provider they are asked to login with a test account. This test account's principal is bound to the local user that was used to configure the provider. This gives the user instant access to the authentication provider's information and allows them to configure other users right away.

If the authentication provider is disabled and the reconfigured with a different test user, the request should be rejected. The only possible solutions for this would involve breaking audit trails or not binding the user to the local account, causing confusion when they attempt to continue with setup. Instead the user should be informed why rancher cannot proceed and given guidance on how to proceed.

**Comment by snasovich** [comment]
This already has PR up for backend but has a dependency on #33331. Once it's done this will be resumed.

**Comment by philomory** [comment]
On a related note, is there some way to avoid having the credentials used to test AD bound as a principal at all? We _do_ have a service account we're using, that exists _only_ to verify LDAP authentication, but because we're having Terraform supply those test credentials when it sets up the AD integration, those dummy credentials (which are widely known) become bound to the Rancher default admin account. Now, I can go in and `kubectl edit` the default admin account afterwards and remove the AD principal, but, this _really_ seems like it shouldn't be necessary.

**Comment by paynejacob** [comment]
@philomory The test account is intended to be the AD account of the admin.  If you are testing with an account that exists only for setting up rancher then that user should be the admin.  If we did not bind the admin account to the test user there would be no way to link the admin account to an AD account.  This would mean there is a user who can login with AD credentials and act as 1 user and be able to login with local credentials and act as another users.

**Comment by vincent99** [comment]
I don't think there's an option in the provider for this currently, but you/terraform don't _need_ to go through the test flow at all.  It is possible to just directly set the auth config without doing a test or associating a principalId to anybody.

**Comment by izaac** [comment]
Rancher version: `v2.6.0`

This is reproducible on the latest Rancher release `v2.6.0`

**Comment by snasovich** [comment]
@izaac , this is expected as neither backend change PR is not merged in yet and necessary UI change is in "next up".

**Comment by davidnuzik** [comment]
@snasovich Sowmya asked QA to try to reproduce this - I do not know specifically why but this is why Izaac mentioned that.

**Comment by snasovich** [comment]
@paynejacob , looks like UI work is done. Could you please get the back-end part through? Let me know if we're blocked on anything else.
